lib/ansible/module_utils/network/ftd/configuration.py

```diff
@@ -28,86 +30,358 @@
-    def __init__(self, conn):
+
+    def __init__(self, conn, check_mode=False):
         self._conn = conn
         self.config_changed = False
+        self._operation_spec_cache = {}
+        self._models_operations_specs_cache = {}
+        self._check_mode = check_mode
+        self._operation_checker = OperationChecker
 
-    def get_object_by_name(self, url_path, name, path_params=None):
-        item_generator = iterate_over_pageable_resource(
-            partial(self.send_request, url_path=url_path, http_method=HTTPMethod.GET, path_params=path_params),
-            {'filter': 'name:%s' % name}
-        )
-        # not all endpoints support filtering so checking name explicitly
-        return next((item for item in item_generator if item['name'] == name), None)
+    def execute_operation(self, op_name, params):
+        """
+        Allow user request execution of simple operations(natively supported by API provider) as well as complex
+        operations(operations that are implemented as a set of simple operations).
+
+        :param op_name: name of the operation being called by the user
+        :type op_name: str
+        :param params: definition of the params that operation should be executed with
+        :type params: dict
+        :return: Result of the operation being executed
+        :rtype: dict
+        """
+        if self._operation_checker.is_upsert_operation(op_name):
+            return self.upsert_object(op_name, params)
+        else:
+            return self.crud_operation(op_name, params)
 
-    def get_objects_by_filter(self, url_path, filters, path_params=None, query_params=None):
-        def match_filters(obj):
-            for k, v in filters.items():
+    def crud_operation(self, op_name, params):
+        """
+        Allow user request execution of simple operations(natively supported by API provider) only.
+
+        :param op_name: name of the operation being called by the user
+        :type op_name: str
+        :param params: definition of the params that operation should be executed with
+        :type params: dict
+        :return: Result of the operation being executed
+        :rtype: dict
+        """
+        op_spec = self.get_operation_spec(op_name)
+        if op_spec is None:
+            raise FtdInvalidOperationNameError(op_name)
+
+        if self._operation_checker.is_add_operation(op_name, op_spec):
+            resp = self.add_object(op_name, params)
+        elif self._operation_checker.is_edit_operation(op_name, op_spec):
+            resp = self.edit_object(op_name, params)
+        elif self._operation_checker.is_delete_operation(op_name, op_spec):
+            resp = self.delete_object(op_name, params)
+        elif self._operation_checker.is_find_by_filter_operation(op_name, params, op_spec):
+            resp = list(self.get_objects_by_filter(op_name, params))
+        else:
+            resp = self.send_general_request(op_name, params)
```

Comment:
I agree that it would be nice to validate the operation names before running the playbook somehow, but at the same time we want to make it easy to use. In the current implementation, a single `operation` field defines both the action (`add`, `edit`, etc) and entity (`NetworkObject`, `AccessRule`, etc).

Replacing the action with an enum having `['get', 'add', 'edit', 'delete', 'upsert', 'other']` values, we will also need to add two extra fields:
- `operation_other` for `other` operations;
- `entity_type` for CRUD operations.

At the moment, replacing one `operation` field with three (`operation_type`, `entity_type` and `operation_other`) sounds complicated from the point of user experience. In the future, we might need to separate non-CRUD operations into a separate module and it will simplify things.

Reply:
If you plan to separate non-CRUD operations in future , then better to do it now ! it will never happen in future because it will always break backward compatibility. Once customer starts using it this way there is going to be much higher resistance to change the API.

Reply:
Sorry for my bad explanation. What I meant is that `ftd_configuration` will remain the primary module for managing FTD configuration. We have already created the [docs](https://developer.cisco.com/site/ftd-ansible) containing the list of operations it supports and some people have already started using it. 
As some non-CRUD operations require multiple steps (tasks) to be executed (e.g., to make a deployment, you start a deployment job first and wait until it finishes), we thought about creating a helper module (or even Ansible roles) to standardize boilerplate code for such common tasks in the future. Though, these tasks can be still achieved by writing multiple/complex plays using `ftd_configuration` module.